### PR TITLE
SAK-23630 new unboundid ldap provider

### PR DESCRIPTION
**Pull request by ottenhoff**
Resurrecting the Unboundid provider code. @marktriggs I pulled two commits from your sakai11 branch. Do you have more improvements we should look at?

### CODE REVIEW

**Comment by marktriggs** [comment]
Hi @ottenhoff,

Another two things we did that might be worth considering...

One was a little one to let us set the connection pool size in sakai.properties:

https://github.com/NYUeServ/sakai11/commit/09553e3d35f1603532833de2d0fd22f29610c3c2

The other was to switch to a SingleServetSet instead of RoundRobinServerSet.  That commit is https://github.com/NYUeServ/sakai11/commit/85723da296c17288ce7134ee81a825be0a5c321f.  Maybe this only makes sense if there's just one LDAP server in play (which is the case for NYU), but here's what I wrote at the time (I think I've shared the attached google doc with the world, but let me know if you can't get it):

> Hi all,
> 
> I've just been looking at the performance of the new LDAP driver we're using (Unboundid).
> 
> To recap what we saw this afternoon (well, my morning): the new code to log slow requests reported a bunch of threads like this:
> 
>  Thread has been running for 130404 ms: http-nio-8080-exec-147::1485806650312.  Stack follows com.unboundid.ldap.sdk.RoundRobinServerSet.getConnection(RoundRobinServerSet.java:311) com.unboundid.ldap.sdk.LDAPConnectionPool.createConnection(LDAPConnectionPool.java:1201) com.unboundid.ldap.sdk.LDAPConnectionPool.createConnection(LDAPConnectionPool.java:1177) com.unboundid.ldap.sdk.LDAPConnectionPool.handleDefunctConnection(LDAPConnectionPool.java:2070) com.unboundid.ldap.sdk.LDAPConnectionPool.releaseDefunctConnection(LDAPConnectionPool.java:2034)
> 
> when things get slow and we start seeing the health check alerts, this is what most of the slow threads are doing: they're waiting for the RoundRobinServerSet to give them a connection to the LDAP server. Looking at the LDAP driver's code for this, it's actually restricting the connection process down to a single thread--so while one thread is making a connection to LDAP, all other threads have to wait for it.
> 
> My guess is that when things go bad, it's because LDAP connections started taking longer than normal. This has the effect of creating a long queue of threads waiting for connections, which slows everything down.
> 
> Looking at the Unboundid docs, it seems that we can switch from a RoundRobinServerSet (which is really intended to spread connections across multiple servers), to a SingleServerSet (which just connects to one server). The SingleServerSet just creates connections on demand (with no pooling), but that's fine: the old driver did that too.
> 
> I ran a little test locally to check the behavior of the two server sets. I added 100ms of artificial network latency, then ran 1000 threads all trying to get a connection as quickly as possible from each of the pool types. You can see the result here:
> 
> https://docs.google.com/a/nyu.edu/spreadsheets/d/1vSrpL1GQvTMWKLAaJbLR8Hknu8OOsXK0US5yn5nyY7s/edit?usp=sharing
> 
> The RoundRobinSet (first sheet; what's currently in production) performs terribly in the presence of latency: you can see the threads form a perfect queue, with the quickest getting a connection in 2.5 seconds, and the slowest getting a connection in just under 6 seconds.
> 
> The same experiment with the SingleServerSet (second sheet) shows basically no contention between the threads, with the majority getting a connection in 400ms, and the worst case taking 800ms.

**Comment by ottenhoff** [comment]
Thanks Mark. I grabbed the first commit. Interesting notes re performance. Do you see anything in the release notes to indicate upstream fixes? 

https://docs.ldap.com/ldap-sdk/docs/release-notes.html

**Comment by marktriggs** [comment]
@ottenhoff I was originally looking at version 3.1.0, so I just grabbed the code for 4.0.7 to have a look.  I think the issue still stands--the `getConnection` method is still marked as synchronized, so I think if latency got bad the connections would still stack up.

NYU has been running SingleServerSet since January 2017 and hasn't had problems, so probably a pretty safe bet.

File: providers/unboundid/pom.xml [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+      <artifactId>jmock-junit4</artifactId>
+      <version>2.8.2</version>
+      <scope>test</scope>
+    </dependency>
+<!-- This exists strictly so we can control logging
+    	levels in Maven-executed unit tests -->
+    <dependency>
+      <groupId>log4j</groupId>
+      <artifactId>log4j</artifactId>
+      <version>1.2.8</version>
```
Comment (ern) [inline providers/unboundid/pom.xml]: should used the inherited version of log4j, i.e. remove version and scope

File: providers/unboundid/src/java/org/sakaiproject/unboundid/EntryContainerRdnToUserTypeMapper.java [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+	 * value mapping.
+	 * 
+	 * @param ldapEntry the user's <code>LDAPEntry</code>
+	 * @param mapper a source of mapping configuration
+	 * @return the entry's mapped RDN
+	 */
+	public String mapLdapEntryToSakaiUserType(LDAPEntry ldapEntry,
+			LdapAttributeMapper mapper) {
+		
+		if ( log.isDebugEnabled() ) {
```
Comment (ern) [inline providers/unboundid/src/java/org/sakaiproject/unboundid/EntryContainerRdnToUserTypeMapper.java]: put this logging line after dnString and use newer slf4j logging

File: providers/unboundid/src/java/org/sakaiproject/unboundid/LdapUserData.java [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+
+/**
+ * Snapshot of a user directory entry, typically used for
+ * cache entries in {@link UnboundidDirectoryProvider}
+ * 
+ * @author David Ross, Albany Medical College
+ * @author Rishi Pande, Virginia Tech
+ * @author Dan McCallum, Unicon Inc
+ */
+public class LdapUserData
```
Comment (ern) [inline providers/unboundid/src/java/org/sakaiproject/unboundid/LdapUserData.java]: Use of lomboks @Data would remove a lot of this bolier plate

File: providers/unboundid/src/java/org/sakaiproject/unboundid/SimpleLdapAttributeMapper.java [hunk 4]
```diff
@@ -0,0 +1,677 @@
...
+	 * </p>
+	 * 
+	 * <p>This defaulting enables UDP config 
+	 * forward-compatibility.</p>
+	 * 
+	 */
+	public void init() {
+		
+		if ( log.isDebugEnabled() ) {
+			log.debug("init()");
```
Comment (ern) [inline providers/unboundid/src/java/org/sakaiproject/unboundid/SimpleLdapAttributeMapper.java]: remove this useless logging line

File: providers/unboundid/src/java/org/sakaiproject/unboundid/SimpleLdapAttributeMapper.java [hunk 5]
```diff
@@ -0,0 +1,677 @@
...
+		}
+		
+		if ( attributeMappings == null || attributeMappings.isEmpty() ) {
+			if ( log.isDebugEnabled() ) {
+				log.debug("init(): creating default attribute mappings");
```
Comment (ern) [inline providers/unboundid/src/java/org/sakaiproject/unboundid/SimpleLdapAttributeMapper.java]: improve logging by removing:
```
if ( log.isDebugEnabled() ) {
	log.debug("init(): creating default attribute mappings");
}
```
and replace with:
```
log.debug("Creating default attribute mappings");
```

File: providers/unboundid/src/java/org/sakaiproject/unboundid/SimpleLdapAttributeMapper.java [hunk 6]
```diff
@@ -0,0 +1,677 @@
...
+			}
+			setAttributeMappings(AttributeMappingConstants.DEFAULT_ATTR_MAPPINGS);
+		}
+		 
+		if ( userTypeMapper == null ) {
+			userTypeMapper = new EmptyStringUserTypeMapper();
+			if ( log.isDebugEnabled() ) {
+				log.debug("init(): created default user type mapper [mapper = " +
```
Comment (ern) [inline providers/unboundid/src/java/org/sakaiproject/unboundid/SimpleLdapAttributeMapper.java]: improve logging here

File: providers/unboundid/src/java/org/sakaiproject/unboundid/SimpleLdapAttributeMapper.java [hunk 7]
```diff
@@ -0,0 +1,677 @@
...
+			userTypeMapper = new EmptyStringUserTypeMapper();
+			if ( log.isDebugEnabled() ) {
+				log.debug("init(): created default user type mapper [mapper = " + 
+						userTypeMapper + "]");
+			}
+		}
+		if ( valueMappings == null ) {
+			valueMappings = Collections.emptyMap();
+			if ( log.isDebugEnabled() ) {
+				log.debug("init(): created default value mapper [mapper = " +
```
Comment (ern) [inline providers/unboundid/src/java/org/sakaiproject/unboundid/SimpleLdapAttributeMapper.java]: improve logging here

File: providers/unboundid/src/java/org/sakaiproject/unboundid/SimpleLdapAttributeMapper.java [hunk 8]
```diff
@@ -0,0 +1,677 @@
...
+     * over {@link LDAPEntry#getAttributeSet()}, handing each attribute
+     * to {@link #mapLdapAttributeOntoUserData(LDAPAttribute, LdapUserData, Collection)}.  
+     * Then enforces the preferred first name field, if it exists.
+     * Finally, assigns a "type" to the {@link LdapUserData} as defined
+     * by {@link #mapLdapEntryToSakaiUserType(LDAPEntry)}.
+	 * 
+     * @see UserTypeMapper
+	 * @param ldapEntry the user's directory entry
+	 * @param userData target {@link LdapUserData}
+	 */
```
Comment (ern) [inline providers/unboundid/src/java/org/sakaiproject/unboundid/SimpleLdapAttributeMapper.java]: spacing

File: providers/unboundid/src/java/org/sakaiproject/unboundid/SimpleLdapAttributeMapper.java [hunk 9]
```diff
@@ -0,0 +1,677 @@
...
+     * @param userData the target {@link LdapUserData} instance
+     * @param logicalAttrNames logical name(s) of the <code>attribute</code>. May
+     *   be null or empty, indicating no configured logical name(s).
+     */
+    protected void mapLdapAttributeOntoUserData(LDAPAttribute attribute, 
+            LdapUserData userData, Collection<String> logicalAttrNames) {
+        
+        if ( logicalAttrNames == null || logicalAttrNames.isEmpty() ) {
+            if ( log.isDebugEnabled() ) {
+                log.debug("No logical name for attribute. [physical name = " +
```
Comment (ern) [inline providers/unboundid/src/java/org/sakaiproject/unboundid/SimpleLdapAttributeMapper.java]: improve logging here

File: providers/unboundid/src/java/org/sakaiproject/unboundid/SimpleLdapAttributeMapper.java [hunk 10]
```diff
@@ -0,0 +1,677 @@
...
+    protected void mapLdapAttributeOntoUserData(LDAPAttribute attribute, 
+            LdapUserData userData, String logicalAttrName) {
+        
+    	Attribute unboundidAttribute = attribute.toAttribute();
+        String attrValue = unboundidAttribute.getValue();
+        MessageFormat format = valueMappings.get(logicalAttrName);
+        if (format != null && attrValue != null) {
+            format = (MessageFormat)format.clone();
+            if ( log.isDebugEnabled() ) {
+                log.debug("mapLdapAttributeOntoUserData(): value mapper [attrValue = " +
```
Comment (ern) [inline providers/unboundid/src/java/org/sakaiproject/unboundid/SimpleLdapAttributeMapper.java]: improve logging here

File: providers/unboundid/src/java/org/sakaiproject/unboundid/SimpleLdapAttributeMapper.java [hunk 11]
```diff
@@ -0,0 +1,677 @@
...
+		for ( Iterator<String> eidIterator = criteria.iterator(); eidIterator.hasNext(); ) {
+			sb.append("(");
+			sb.append(getFindUserByEidFilter(eidIterator.next()));
+			sb.append(")");
+		}
+		
+		sb.append(")");
+		
+		if (log.isDebugEnabled()) {
+			log.debug("getManyUsersInOneSearch() completed filter: " + sb.toString());
```
Comment (ern) [inline providers/unboundid/src/java/org/sakaiproject/unboundid/SimpleLdapAttributeMapper.java]: improve logging here

File: providers/unboundid/src/java/org/sakaiproject/unboundid/StringUserTypeMapper.java [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+ * Very simple {@link UserTypeMapper} which generates the same
+ * <code>String</code> value for any users. Technically, this
+ * object could be reconfigured at any time.
+ * 
+ * @author dmccallum
+ */
+@Slf4j
+public class StringUserTypeMapper implements UserTypeMapper {
+	
+	private String userType;
```
Comment (ern) [inline providers/unboundid/src/java/org/sakaiproject/unboundid/StringUserTypeMapper.java]: lombok @Getter @Setter

File: providers/unboundid/src/java/org/sakaiproject/unboundid/UnboundidDirectoryProvider.java [hunk 13]
```diff
@@ -0,0 +1,1516 @@
...
+	}
+
+	/**
+	 * Typically called by Spring to signal bean destruction.
+	 *
+	 */
+	public void destroy()
+	{
+		if ( log.isDebugEnabled() ) {
+			log.debug("destroy()");
```
Comment (ern) [inline providers/unboundid/src/java/org/sakaiproject/unboundid/UnboundidDirectoryProvider.java]: useless logging

File: providers/unboundid/pom.xml [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+    <dependency>
+      <groupId>org.zapodot</groupId>
+      <artifactId>embedded-ldap-junit</artifactId>
+      <version>0.7</version>
+      <scope>test</scope>
+    </dependency>
+  </dependencies>
+  <build>
+    <plugins>
+      <plugin>
```
Comment (ern) [inline providers/unboundid/pom.xml]: maven-surefire-plugin is already declared in master probably not needed.

**Comment by smarquard** [comment]
I've tested this in our 12.x dev server against our MS AD LDAP server, and the basics work as expected (LDAP search, authenticate), configured much the same way as the jldap provider.

**Comment by ottenhoff** [comment]
Alright @ern made all your logging changes ;)

**Pull request merged**